libp2p/protocols/mix/utils.nim

```diff
@@ -0,0 +1,147 @@
+  ):
+    return err("Invalid multiaddress format")
+
+  if parts[1] != "ip4":
+    # ToDo: Add support for ipv6. Supporting ipv4 only for testing purposes
+    return err("Only ip4 addresses are supported")
+
+  # IP address (4 bytes) 
+  let ipParts = parts[2].split('.')
+  if ipParts.len != 4:
+    return err("Invalid IP address format")
+  for part in ipParts:
+    try:
+      let ipPart = parseInt(part)
+      if ipPart < 0 or ipPart > 255:
+        return err("Invalid IP address format")
+      res.add(byte(ipPart))
+    except ValueError:
+      return err("Invalid IP address format")
+
+  # Protocol (1 byte) ToDo: TLS or QUIC
+  var isQuic = false
+  if parts[3] != "tcp" and not (parts[3] == "udp" and parts[5] == "quic-v1"):
+    return err("Unsupported protocol")
+  res.add(
+    if parts[3] == "tcp":
+      byte(0)
+    else:
+      isQuic = true
+      byte(1)
+  ) # Using TCP for testing purposes
+
+  # Port (2 bytes)
+  try:
+    let port = parseInt(parts[4])
+    if port < 0 or port > 65535:
+      return err("Invalid port")
+    res.add(uint16ToBytes(uint16(port)))
+  except ValueError:
+    return err("Invalid port")
+
+  # PeerID (39 bytes),  if using circuit relay, this represents the relay server
+  let peerId1Bytes = ?extractPeerId(parts, if isQuic: 7 else: 6)
+  res.add(peerId1Bytes)
+
+  if parts.len > minMultiAddrComponentLen + (if isQuic: 1 else: 0):
+    if not (
+      parts.len == maxMultiAddrComponentLen - 1 or
+      (isQuic and parts.len == maxMultiAddrComponentLen)
+    ):
+      return err("invalid p2p circuit relay multiaddress")
+    elif not (parts[if isQuic: 8 else: 7] == "p2p-circuit"):
+      return err("Invalid p2p-circuit component")
+
+    let peerId2Bytes = ?extractPeerId(parts, if isQuic: 10 else: 9)
+    res.add(peerId2Bytes)
+
+  if res.len > addrSize:
+    return err("Address must be <= " & $addrSize & " bytes")
+
+  return ok(res & newSeq[byte](addrSize - res.len))
+
+proc bytesToMultiAddr*(bytes: openArray[byte]): Result[string, string] =
+  if bytes.len != addrSize:
+    return err("Address must be exactly " & $addrSize & " bytes")
+
+  var ipParts: seq[string] = @[]
+  for i in 0 .. 3:
+    ipParts.add($bytes[i])
+      # ToDo: Add support for ipv6. Supporting ipv4 only for testing purposes
+
+  let protocol = if bytes[4] == 0: "tcp" else: "udp"
+
+  let quic = if bytes[4] == 1: "/quic-v1" else: ""
+
+    # ToDo: TLS or QUIC (Using TCP for testing purposes)
+
+  let port = bytesToUInt16(bytes[5 .. 6]).valueOr:
+    return err("Error in conversion of bytes to port no.: " & error)
+
+  let peerId1 = "/p2p/" & Base58.encode(bytes[7 ..< 46])
+
+  let peerId2Bytes = bytes[7 + peerIdByteLen ..< 7 + (peerIdByteLen * 2)]
+  let peerId2 =
+    if peerId2Bytes != newSeq[byte](peerIdByteLen):
+      "/p2p-circuit/p2p/" & Base58.encode(peerId2Bytes)
+    else:
+      ""
+
+  return ok(
+    "/ip4/" & ipParts.join(".") & "/" & protocol & "/" & $port & quic & peerId1 & peerId2
+  )
```

Comment:
If we don't have these functions in `multiaddress.nim`, they should be included there and then used here, not defined here.

Reply:
While the code can use some of the code from multiaddress.nim, the serialization/deserialization functions are specific for mix protocol, as defined in https://github.com/vacp2p/rfc-index/blob/cc8c7a0ac923d382e5ca778ffa44dbc21c80ec33/vac/raw/mix.md#84-address-and-delay-encoding . I'll still change them a bit because I think these functions should live maybe in serialization.nim from mix, and also use functions from multiaddress.nim to determine whether components from multiaddress exist or not.